File: rudder-client-ios/RudderClient/Event/Properties/PagePropertyBuilder.swift
//
//  PagePropertyBuilder.swift
//  RudderSample
//
//  Created by Arnab Pal on 17/07/19.
//  Copyright © 2019 Arnab Pal. All rights reserved.
//
import Foundation

class PagePropertyBuilder : RudderPropertyBuilder {
    private var title: String = ""
    func setTitle(title: String) -> PagePropertyBuilder {
        self.title = title
        return self
    }
    
    private var url: String? = nil
    func setUrl(url: String) -> PagePropertyBuilder {
        self.url = url
        return self
    }
    
    private var path: String = ""
    func setPath(path: String) -> PagePropertyBuilder {
        self.path = path
        return self
    }
    
    private var referrer: String = ""
    func setReferrer(referrer: String) -> PagePropertyBuilder {
        self.referrer = referrer
        return self
    }
    
    private var search: String = ""
    func setSearch(search: String) -> PagePropertyBuilder {
        self.search = search
        return self
    }
    
    private var keywords: String = ""
    func setKeywords(keywords: String) -> PagePropertyBuilder {
        self.keywords = keywords
        return self
    }
    
    func buildProperty() throws -> RudderProperty {
        if (self.url == nil){
            throw MalformedEventError(_message: "Key \"url\" is required for page event")
        }
        return try self.addStringProperty(key: "title", value: self.title)
            .addStringProperty(key: "url", value: self.url!)
            .addStringProperty(key: "path", value: self.path)
            .addStringProperty(key: "referrer", value: self.referrer)
            .addStringProperty(key: "search", value: self.search)
            .addStringProperty(key: "keywords", value: self.keywords)
            .build()
    }
}

File: old-references/rudder-client-android-kotlin-old/src/main/java/com/rudderlabs/android/library/models/porperties/PagePropertyBuilder.kt
package com.rudderlabs.android.library.models.porperties

import com.rudderlabs.android.library.exceptions.MalformedEventException

class PagePropertyBuilder : RudderPropertyBuilder() {
    private var title: String = ""
    fun setTitle(title: String): PagePropertyBuilder {
        this.title = title
        return this
    }

    private var url: String? = null
    fun setUrl(url: String): PagePropertyBuilder {
        this.url = url
        return this
    }

    private var path: String = ""
    fun setPath(path: String): PagePropertyBuilder {
        this.path = path
        return this
    }

    private var referrer: String = ""
    fun setReferrer(referrer: String): PagePropertyBuilder {
        this.referrer = referrer
        return this
    }

    private var search: String = ""
    fun setSearch(search: String): PagePropertyBuilder {
        this.search = search
        return this
    }

    private var keywords: String = ""
    fun setKeywords(keywords: String): PagePropertyBuilder {
        this.keywords = keywords
        return this
    }

    override fun build(): RudderProperty? {
        if (url == null) {
            throw MalformedEventException("Key \"url\" is required for track event")
        }
        return RudderProperty().also {
            it.addProperty("title", this.title)
            it.addProperty("url", this.url)
            it.addProperty("path", this.path)
            it.addProperty("referrer", this.referrer)
            it.addProperty("search", this.search)
            it.addProperty("keywords", this.keywords)
        }
    }
}